### PR TITLE
Improve doc for auto compaction

### PR DESCRIPTION
**Pull request by jihoonson**

### CODE REVIEW

File: docs/content/design/coordinator.md [hunk 1]
```diff
@@ -86,17 +86,30 @@ Once a compact task fails, the coordinator simply finds the segments for the int
...
-The segment name indicates its dataSource and interval. The search result of newestSegmentFirstPolicy is [`seg_ds3_2017-12-01_2017-12-02`, `seg_ds1_2017-11-01_2017-11-02`, `seg_ds1_2017-10-01_2017-10-02`, `seg_ds2_2017-08-01_2017-08-02`, `seg_ds3_2017-07-01_2017-07-02`].
-
-Every run, this policy starts searching from the (very latest interval - [skipOffsetFromLatest](../configuration/index.html#compaction-dynamic-configuration)).
-This is to handle the late segments ingested to realtime dataSources.
+At every coordinator run, this policy searches segments to compact by iterating segments from the latest to the oldest.
+Once it finds the latest segment among all dataSources, it checks the segment is _compactible_ with other segments of the same dataSource which have the same or abutting intervals.
+Note that segments are compactible if their total size is smaller than or equal to the configured `inputSegmentSizeBytes`.
+
+Let me give you some more details with an example. Let me assume we have two dataSources (`foo`, `bar`)
```
Comment (nishantmonu51) [inline docs/content/design/coordinator.md]: Can we rephrase as - 
"Here are some details with example, Let us assume..."

File: docs/content/design/coordinator.md [hunk 2]
```diff
@@ -86,17 +86,30 @@ Once a compact task fails, the coordinator simply finds the segments for the int
...
+When `inputSegmentSizeBytes` is 20 MB, this policy first returns two segments (`foo_2017-11-01/2017-12-01` and `foo_2017-10-01/2017-11-01`) to compact because
+they are the latest segment and its abutting segment, and their total size is equal to `inputSegmentSizeBytes`.
+
+If the coordinator has enough task slots for compaction, this policy would continue searching the next segments and return
+`bar_2017-10-01/2017-11-01` and `bar_2017-09-01/2017-10-01`.
+Note that `bar_2017-08-01/2017-09-01` is not compacted together even though it abuts to `bar_2017-09-01/2017-10-01`.
+This is because the total segment size to compact would be greater than `inputSegmentSizeBytes` if it's included.
+
+The search start point can be changed by setting [skipOffsetFromLatest](../configuration/index.html#compaction-dynamic-configuration)).
+If this is set, this policy will ignore the segments failling into the interval of (the end time of the very latest segment - `skipOffsetFromLatest`).
```
Comment (nishantmonu51) [inline docs/content/design/coordinator.md]: s/failling/falling/g

File: docs/content/design/coordinator.md (hunk 1 above)
Comment (jihoonson) [reply docs/content/design/coordinator.md]: Sounds good. Fixed.

File: docs/content/design/coordinator.md (hunk 2 above)
Comment (jihoonson) [reply docs/content/design/coordinator.md]: Thanks!

**Comment by jihoonson** [comment]
@nishantmonu51 thanks for the review. I've addressed your comments.

File: docs/content/configuration/index.md [hunk 3]
```diff
@@ -806,7 +806,7 @@ A description of the compaction config is:
 |`keepSegmentGranularity`|Set [keepSegmentGranularity](../ingestion/compaction.html) to true for compactionTask.|no (default = true)|
 |`taskPriority`|[Priority](../ingestion/tasks.html#task-priorities) of compact task.|no (default = 25)|
 |`inputSegmentSizeBytes`|Total input segment size of a compactionTask.|no (default = 419430400)|
```
Comment (gianm) [inline docs/content/configuration/index.md]: How about:

> Maximum number of total segment bytes processed per compaction task. Since a time chunk must be processed in its entirety, if the segments for a particular time chunk have a total size in bytes greater than this parameter, compaction will not run for that time chunk. Because each compaction task runs with a single thread, setting this value too far above 1–2GB will result in compaction tasks taking an excessive amount of time.

File: docs/content/configuration/index.md [hunk 4]
```diff
@@ -806,7 +806,7 @@ A description of the compaction config is:
...
-|`targetCompactionSizeBytes`|The target segment size of compaction. The actual size of a compact segment might be slightly larger or smaller than this value.|no (default = 419430400)|
+|`targetCompactionSizeBytes`|The target segment size after compaction. The actual size of a compact segment might be slightly larger or smaller than this value.|no (default = 419430400)|
```
Comment (gianm) [inline docs/content/configuration/index.md]: How about:

> The target segment size, for each segment, after compaction. The actual sizes of compacted segments might be slightly larger or smaller than this value. Each compaction task may generate more than one output segment, and it will try to keep each output segment close to this configured size.

File: docs/content/configuration/index.md [hunk 5]
```diff
@@ -806,7 +806,7 @@ A description of the compaction config is:
...
 |`maxNumSegmentsToCompact`|Max number of segments to compact together.|no (default = 150)|
```
Comment (gianm) [inline docs/content/configuration/index.md]: How about:

> Maximum number of segments to compact together per compaction task. Since a time chunk must be processed in its entirety, if a time chunk has a total number of segments greater than this parameter, compaction will not run for that time chunk.

File: docs/content/configuration/index.md [hunk 6]
```diff
@@ -820,7 +820,12 @@ An example of compaction config is:
 }
 ```
 
-For realtime dataSources, it's recommended to set `skipOffsetFromLatest` to some sufficiently large value to avoid frequent compact task failures.
+Note that compact tasks can fail if their locks are revoked by other tasks of higher priorities.
+Since realtime tasks have a higher priority than compact task by default,
```
Comment (gianm) [inline docs/content/configuration/index.md]: compact tasks (grammar)

File: docs/content/configuration/index.md [hunk 7]
```diff
@@ -820,7 +820,12 @@ An example of compaction config is:
...
+it can be problematic if there're frequent conflicts between compact tasks and realtime tasks.
```
Comment (gianm) [inline docs/content/configuration/index.md]: there are (too colloquial)

File: docs/content/configuration/index.md [hunk 8]
```diff
@@ -820,7 +820,12 @@ An example of compaction config is:
...
+If this is the case, coordinator's automatic compaction might be stuck because of the frequent compact task fails.
```
Comment (gianm) [inline docs/content/configuration/index.md]: the coordinator's

Comment (gianm) [reply docs/content/configuration/index.md]: might get stuck

because of frequent compact task failures

File: docs/content/design/coordinator.md [hunk 9]
```diff
@@ -86,17 +86,30 @@ Once a compact task fails, the coordinator simply finds the segments for the int
 
 #### Newest Segment First Policy
 
-This policy searches the segments of _all dataSources_ in inverse order of their intervals.
-For example, let me assume there are 3 dataSources (`ds1`, `ds2`, `ds3`) and 5 segments (`seg_ds1_2017-10-01_2017-10-02`, `seg_ds1_2017-11-01_2017-11-02`, `seg_ds2_2017-08-01_2017-08-02`, `seg_ds3_2017-07-01_2017-07-02`, `seg_ds3_2017-12-01_2017-12-02`) for those dataSources.
-The segment name indicates its dataSource and interval. The search result of newestSegmentFirstPolicy is [`seg_ds3_2017-12-01_2017-12-02`, `seg_ds1_2017-11-01_2017-11-02`, `seg_ds1_2017-10-01_2017-10-02`, `seg_ds2_2017-08-01_2017-08-02`, `seg_ds3_2017-07-01_2017-07-02`].
-
-Every run, this policy starts searching from the (very latest interval - [skipOffsetFromLatest](../configuration/index.html#compaction-dynamic-configuration)).
-This is to handle the late segments ingested to realtime dataSources.
+At every coordinator run, this policy searches segments to compact by iterating segments from the latest to the oldest.
```
Comment (gianm) [inline docs/content/design/coordinator.md]: searches for

File: docs/content/design/coordinator.md [hunk 10]
```diff
@@ -86,17 +86,30 @@ Once a compact task fails, the coordinator simply finds the segments for the int
...
+Once it finds the latest segment among all dataSources, it checks the segment is _compactible_ with other segments of the same dataSource which have the same or abutting intervals.
```
Comment (gianm) [inline docs/content/design/coordinator.md]: if the segment

File: docs/content/design/coordinator.md [hunk 11]
```diff
@@ -86,17 +86,30 @@ Once a compact task fails, the coordinator simply finds the segments for the int
...
+Note that segments are compactible if their total size is smaller than or equal to the configured `inputSegmentSizeBytes`.
+
+Here are some details with example. Let us assume we have two dataSources (`foo`, `bar`)
```
Comment (gianm) [inline docs/content/design/coordinator.md]: an example

File: docs/content/design/coordinator.md [hunk 12]
```diff
@@ -86,17 +86,30 @@ Once a compact task fails, the coordinator simply finds the segments for the int
...
+and 5 segments (`foo_2017-10-01/2017-11-01`, `foo_2017-11-01/2017-12-01`, `bar_2017-08-01/2017-09-01`, `bar_2017-09-01/2017-10-01`, `bar_2017-10-01/2017-11-01`).
```
Comment (gianm) [inline docs/content/design/coordinator.md]: Would be better for these segments to have a normal looking segment id

Comment (gianm) [reply docs/content/design/coordinator.md]: Or maybe just leave the version off but use a `_` so it's like a prefix of a normal segment id

File: docs/content/design/coordinator.md [hunk 13]
```diff
@@ -86,17 +86,30 @@ Once a compact task fails, the coordinator simply finds the segments for the int
...
+The segment name indicates its dataSource and interval, and each segment has the same size of 10 MB.
+When `inputSegmentSizeBytes` is 20 MB, this policy first returns two segments (`foo_2017-11-01/2017-12-01` and `foo_2017-10-01/2017-11-01`) to compact because
+they are the latest segment and its abutting segment, and their total size is equal to `inputSegmentSizeBytes`.
+
+If the coordinator has enough task slots for compaction, this policy would continue searching the next segments and return
```
Comment (gianm) [inline docs/content/design/coordinator.md]: searching for

File: docs/content/configuration/index.md (hunk 3 above)
Comment (jihoonson) [reply docs/content/configuration/index.md]: Thanks!

File: docs/content/configuration/index.md (hunk 4 above)
Comment (jihoonson) [reply docs/content/configuration/index.md]: Thanks!

File: docs/content/configuration/index.md (hunk 5 above)
Comment (jihoonson) [reply docs/content/configuration/index.md]: Thanks!

File: docs/content/configuration/index.md (hunk 6 above)
Comment (jihoonson) [reply docs/content/configuration/index.md]: Thanks!

File: docs/content/configuration/index.md (hunk 7 above)
Comment (jihoonson) [reply docs/content/configuration/index.md]: Fixed.

File: docs/content/configuration/index.md (hunk 8 above)
Comment (jihoonson) [reply docs/content/configuration/index.md]: Thanks!

File: docs/content/design/coordinator.md (hunk 9 above)
Comment (jihoonson) [reply docs/content/design/coordinator.md]: Thanks!

File: docs/content/design/coordinator.md (hunk 10 above)
Comment (jihoonson) [reply docs/content/design/coordinator.md]: Thanks!

File: docs/content/design/coordinator.md (hunk 11 above)
Comment (jihoonson) [reply docs/content/design/coordinator.md]: Thanks!

File: docs/content/design/coordinator.md (hunk 12 above)
Comment (jihoonson) [reply docs/content/design/coordinator.md]: Changed to something like `foo_2017-10-01T00:00:00.000Z_2017-11-01T00:00:00.000Z_VERSION`. Do you think it's better to use a real version string?

File: docs/content/design/coordinator.md (hunk 13 above)
Comment (jihoonson) [reply docs/content/design/coordinator.md]: Thanks!

**Comment by jihoonson** [comment]
@gianm thanks. I also changed "compact task" to "compaction task" for consistency.

File: docs/content/design/coordinator.md [hunk 14]
```diff
@@ -69,34 +69,46 @@ Each run, the Druid coordinator compacts small segments abutting each other. Thi
...
-The maximum number of running compact tasks is `min(sum of worker capacity * slotRatio, maxSlots)`.
-Note that even though `min(sum of worker capacity * slotRatio, maxSlots)` = 0, at least one compact task is always submitted
+Once some segments are found, it launches a [compaction task](../ingestion/tasks.html#compaction-task) to compact those segments.
+The maximum number of running compaction tasks is `min(sum of worker capacity * slotRatio, maxSlots)`.
+Note that even though `min(sum of worker capacity * slotRatio, maxSlots)` = 0, at least one compaction task is always submitted
 if the compaction is enabled for a dataSource.
 See [Compaction Configuration API](../operations/api-reference.html#compaction-configuration) and [Compaction Configuration](../configuration/index.html#compaction-dynamic-configuration) to enable the compaction.
 
-Compact tasks might fail due to some reasons.
+Compaction tasks might fail due to some reasons.
```
Comment (jon-wei) [inline docs/content/design/coordinator.md]: Suggest rewording this to "Compaction tasks might fail due to the following reasons:"

File: docs/content/design/coordinator.md [hunk 15]
```diff
@@ -69,34 +69,46 @@ Each run, the Druid coordinator compacts small segments abutting each other. Thi
...
+
+If the coordinator has enough task slots for compaction, this policy would continue searching for the next segments and return
+`bar_2017-10-01T00:00:00.000Z_2017-11-01T00:00:00.000Z_VERSION` and `bar_2017-09-01T00:00:00.000Z_2017-10-01T00:00:00.000Z_VERSION`.
+Note that `bar_2017-08-01T00:00:00.000Z_2017-09-01T00:00:00.000Z_VERSION` is not compacted together even though it abuts to `bar_2017-09-01T00:00:00.000Z_2017-10-01T00:00:00.000Z_VERSION`.
+This is because the total segment size to compact would be greater than `inputSegmentSizeBytes` if it's included.
+
+The search start point can be changed by setting [skipOffsetFromLatest](../configuration/index.html#compaction-dynamic-configuration).
+If this is set, this policy will ignore the segments falling into the interval of (the end time of the very latest segment - `skipOffsetFromLatest`).
+This is to avoid conflicts between compaction tasks and realtime tasks.
+Note that realtime tasks have a higher priority than compaction tasks by default. They would revoke compaction tasks if their intervals overlap each other.
```
Comment (jon-wei) [inline docs/content/design/coordinator.md]: Suggest rewording to "Note that realtime tasks have a higher priority than compaction tasks by default. Realtime tasks will revoke the locks of compaction tasks if their intervals overlap, resulting in the termination of the compaction task."

**Comment by jihoonson** [comment]
@jon-wei thanks. Updated.
